### ISSUE TITLE
MigrationQueueDeleteTTLIndex throws an exception

### ISSUE THREAD

**Issue by AndreaFrancis** [huggingface/datasets-server#1326]
Context: https://github.com/huggingface/datasets-server/pull/1238#discussion_r1202804263

While running migrations of type MigrationQueueDeleteTTLIndex, an error is thrown:
`pymongo.errors.OperationFailure: An equivalent index already exists with the same name but different options. Requested index: { v: 2, key: { finished_at: 1 }, name: "finished_at_1", background: false, expireAfterSeconds: 600 }, existing index: { v: 2, key: { finished_at: 1 }, name: "finished_at_1", background: false, expireAfterSeconds: 86400 }, full error: {'ok': 0.0, 'errmsg': 'An equivalent index already exists with the same name but different options. Requested index: { v: 2, key: { finished_at: 1 }, name: "finished_at_1", background: false, expireAfterSeconds: 600 }, existing index: { v: 2, key: { finished_at: 1 }, name: "finished_at_1", background: false, expireAfterSeconds: 86400 }', 'code': 85, 'codeName': 'IndexOptionsConflict', '$clusterTime': {'clusterTime': Timestamp(1684856718, 127), 'signature': {'hash': b'\x9e\x18\x11\xa9\xfd\x9c\xec\x12\x19\xeb\xa7\xad\x94\xfb\x86<Y\xdc\xb4n', 'keyId': 7184422908209397768}}, 'operationTime': Timestamp(1684856718, 127)}`

See also https://huggingface.slack.com/archives/C04L6P8KNQ5/p1684856801355259?thread_ts=1684855715.011289&cid=C04L6P8KNQ5 (internal) for details.

Maybe this is related to other k8s objects running at the same time of deploy and try to re create the old index. Should we stop all pods while deploying? Maybe scale to 0 replicas?

**Comment by AndreaFrancis** [comment]
I think it is probably that we are calling the `mongoengine` Job Document  each time we do a [validation](https://github.com/huggingface/datasets-server/blob/main/jobs/mongodb_migration/src/mongodb_migration/renaming_migrations.py#LL114C17-L114C17) in our migrations `check_documents(DocCls=Job, sample_size=10)`
We should probably use `pymongo` even in validations to avoid schema syncing operations at `mongodb_migration` project level.
I mean, we should probably avoid totally using `mongoengine` in `mongodb_migration` project to handle the db schema layer as it is phisically.

**Comment by severo** [comment]
> Maybe this is related to other k8s objects running at the same time of deploy and try to re create the old index. Should we stop all pods while deploying? Maybe scale to 0 replicas?

No, we don't want to stop the API

**Comment by severo** [comment]
> I mean, we should probably avoid totally using mongoengine (That contains the desired schema) in mongodb_migration project and handle everything with pymongo (The real db schema layer as it is).

I don't think so. It's useful to check if the migration worked and if we can use the database with mongoengine after it

**Comment by severo** [comment]
A workaround could be (https://github.com/huggingface/datasets-server/pull/1325#discussion_r1222660782)

> maybe we have to create a new index (have both in parallel)
> then delete the old one in another migration, in a new deployment (so that no running app would try to recreate it)

**Comment by AndreaFrancis** [comment]
From mongo doc, I don't think it is possible to have two TTL parallel indexes for a field at the same time [Restrictions](https://www.mongodb.com/docs/manual/core/index-ttl/#restrictions). I also tried adding a new index with a different name but it gives the same error.
I think the best/clean way to treat these scenarios is:
1. Delete index code from queue.py and deploy (db physical index will still exist but no API/worker/backfill code will try to create it). 2. Add index delete migration and at the same time we can add the new index value in queue.py, deploy.

It looks like the same as we were doing for previous deploys 
https://github.com/huggingface/datasets-server/pull/1238 and https://github.com/huggingface/datasets-server/pull/1105 (Delete index manually and then run migration) but to standardize we could do it from the code.
Dont think we should overthink a solution for this issue but maybe to have a clear deployment process when doing modifications to the TTL indexes.
wdyt?

**Comment by severo** [comment]
Yes, OK! Two deployments...

**Comment by severo** [comment]
Closing as "won't fix"
